File: solana_toolbox_web3/src/ToolboxIdlEvent.ts
import { ToolboxIdlTypedef } from './ToolboxIdlTypedef';
import { ToolboxIdlTypeFlat } from './ToolboxIdlTypeFlat';
import { hydrate } from './ToolboxIdlTypeFlat.hydrate';
import { parse, parseObjectIsPossible } from './ToolboxIdlTypeFlat.parse';
import { ToolboxIdlTypeFull } from './ToolboxIdlTypeFull';
import { deserialize } from './ToolboxIdlTypeFull.deserialize';
import { serialize } from './ToolboxIdlTypeFull.serialize';
import { ToolboxUtils } from './ToolboxUtils';

export class ToolboxIdlEvent {
  public readonly name: string;
  public readonly docs: any;
  public readonly discriminator: Buffer;
  public readonly infoTypeFlat: ToolboxIdlTypeFlat;
  public readonly infoTypeFull: ToolboxIdlTypeFull;

  constructor(value: {
    name: string;
    docs: any;
    discriminator: Buffer;
    infoTypeFlat: ToolboxIdlTypeFlat;
    infoTypeFull: ToolboxIdlTypeFull;
  }) {
    this.name = value.name;
    this.docs = value.docs;
    this.discriminator = value.discriminator;
    this.infoTypeFlat = value.infoTypeFlat;
    this.infoTypeFull = value.infoTypeFull;
  }

  public static tryParse(
    idlEventName: string,
    idlEvent: any,
    typedefs: Map<string, ToolboxIdlTypedef>,
  ): ToolboxIdlEvent {
    ToolboxUtils.expectObject(idlEvent);
    let docs = idlEvent['docs'];
    let discriminator = Buffer.from(
      ToolboxUtils.expectArray(
        idlEvent['discriminator'] ??
          ToolboxUtils.discriminator(`event:${idlEventName}`),
      ),
    );
    let infoTypeFlat = parseObjectIsPossible(idlEvent)
      ? parse(idlEvent)
      : parse(idlEventName);
    let infoTypeFull = hydrate(infoTypeFlat, new Map(), typedefs);
    return new ToolboxIdlEvent({
      name: idlEventName,
      docs,
      discriminator,
      infoTypeFlat,
      infoTypeFull,
    });
  }

  public encode(eventState: any): Buffer {
    let data: Buffer[] = [];
    data.push(this.discriminator);
    serialize(this.infoTypeFull, eventState, data, true);
    return Buffer.concat(data);
  }

  public decode(eventData: Buffer): any {
    this.check(eventData);
    let [, eventState] = deserialize(
      this.infoTypeFull,
      eventData,
      this.discriminator.length,
    );
    return eventState;
  }

  public check(eventData: Buffer) {
    if (eventData.length < this.discriminator.length) {
      throw new Error('Invalid discriminator');
    }
    for (let i = 0; i < this.discriminator.length; i++) {
      if (eventData[i] !== this.discriminator[i]) {
        throw new Error('Invalid discriminator');
      }
    }
  }
}

File: solana_toolbox_idl_core/src/toolbox_idl_event.rs
use anyhow::anyhow;
use anyhow::Result;
use serde_json::Value;

use crate::toolbox_idl_type_flat::ToolboxIdlTypeFlat;
use crate::toolbox_idl_type_full::ToolboxIdlTypeFull;

#[derive(Debug, Clone, PartialEq)]
pub struct ToolboxIdlEvent {
    pub name: String,
    pub docs: Option<Value>,
    pub discriminator: Vec<u8>,
    pub info_type_flat: ToolboxIdlTypeFlat,
    pub info_type_full: ToolboxIdlTypeFull,
}

impl Default for ToolboxIdlEvent {
    fn default() -> ToolboxIdlEvent {
        ToolboxIdlEvent {
            name: "Unknown".to_string(),
            docs: None,
            discriminator: vec![],
            info_type_flat: ToolboxIdlTypeFlat::nothing(),
            info_type_full: ToolboxIdlTypeFull::nothing(),
        }
    }
}

impl ToolboxIdlEvent {
    pub fn encode(&self, event_message: &Value) -> Result<Vec<u8>> {
        let mut event_data = vec![];
        event_data.extend_from_slice(&self.discriminator);
        self.info_type_full.try_serialize(
            event_message,
            &mut event_data,
            true,
        )?;
        Ok(event_data)
    }

    pub fn decode(&self, event_data: &[u8]) -> Result<Value> {
        self.check(event_data)?;
        let (_, event_value) = self
            .info_type_full
            .try_deserialize(event_data, self.discriminator.len())?;
        Ok(event_value)
    }

    pub fn check(&self, event_data: &[u8]) -> Result<()> {
        if !event_data.starts_with(&self.discriminator) {
            return Err(anyhow!(
                "Invalid event discriminator, expected: {:?}, found: {:?}",
                self.discriminator,
                event_data,
            ));
        }
        Ok(())
    }
}
